cdc/capture_info_test.go

```diff
@@ -93,46 +103,64 @@ func (ci *captureInfoSuite) TestWatch(c *check.C) {
 	watchCtx, watchCancel := context.WithCancel(ctx)
 	infos, watchC, err := newCaptureInfoWatch(watchCtx, ci.client)
 	c.Assert(err, check.IsNil)
+	owner.captureWatchC = watchC
 	// infos contains info1
 	c.Assert(infos, check.HasLen, 1)
 	c.Assert(infos[0], check.DeepEquals, info1)
 
 	mustGetResp := func() *CaptureInfoWatchResp {
 		select {
-		case resp, ok := <-watchC:
-			c.Assert(ok, check.IsTrue)
+		case resp := <-owner.captureWatchC:
+			err := resp.Err
+			if err != nil && errors.Cause(err) == mvcc.ErrCompacted {
+				atomic.AddInt64(&watcherRetry, 1)
+				err2 := owner.resetCaptureInfoWatcher(watchCtx)
+				c.Assert(err2, check.IsNil)
+				return nil
+			}
+			c.Assert(err, check.IsNil)
+			if resp.IsDelete {
+				owner.removeCapture(resp.Info)
+			} else {
+				owner.addCapture(resp.Info)
+			}
 			return resp
-		case <-time.After(time.Second * 5):
+		case <-time.After(time.Second * 2):
 			c.Fatal("timeout to get resp from watchC")
 			return nil
 		}
 	}
 
 	mustClosed := func() {
 		select {
-		case _, ok := <-watchC:
+		case _, ok := <-owner.captureWatchC:
 			c.Assert(ok, check.IsFalse)
-		case <-time.After(time.Second * 5):
+		case <-time.After(time.Second * 2):
 			c.Fatal("timeout to get resp from watchC")
-
 		}
 	}
 
-	// put info2 and info3
+	checkCaptureLen := func(expected int) {
+		owner.l.RLock()
+		defer owner.l.RUnlock()
+		c.Assert(len(owner.captures), check.Equals, expected)
```

Comment:
Can use `HasLen` for better error message?

Reply:
fixed